tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -526,6 +526,21 @@ void promoteToSharedGreedy(
     scop.insertSyncsAroundCopies(bandNode);
   }
 }
+
+// Returns a union of all mapping filters to "MappingType" in "scop".
+template <typename MappingType>
+isl::union_set collectMappingsTo(const Scop& scop) {
+  auto root = scop.scheduleRoot();
+  auto domain = scop.domain();
+  auto mappingFilters = detail::ScheduleTree::collect(
+      root, detail::ScheduleTreeType::MappingFilter);
+  mappingFilters = functional::Filter(isMappingTo<MappingType>, mappingFilters);
+  for (auto mf : mappingFilters) {
+    auto filterNode = mf->elemAs<detail::ScheduleTreeElemMappingFilter>();
+    domain = domain.intersect(filterNode->filter_);
```

Comment:
Perhaps I'm missing something, but this doesn't seem to make sense.
You're looking at multiple mapping filters of the same type and then intersecting them.
Since multiple (thread) mapping filters have disjoint domains, the result will be empty if there is any branching.
I'll request changes for this one.